texar/modules/connectors/connectors.py

```diff
@@ -0,0 +1,786 @@
+            encoder = UnidirectionalRNNEncoder(...)
+            _, final_state = encoder(inputs=...)
+            decoder = BasicRNNDecoder(...)
+            connector = MLPTransformConnector(decoder.state_size)
+            _ = decoder(
+                initial_state=connector(final_state),
+                ...)
+
+    Args:
+        output_size: Size of output **excluding** the batch dimension. For
+            example, set :attr:`output_size` to ``dim`` to generate output of
+            shape ``[batch_size, dim]``.
+            Can be an ``int``, a ``tuple`` of ``int``, a ``torch.Size``,
+            or a ``tuple`` of ``torch.Size``.
+            For example, to transform inputs to have decoder state size, set
+            :python:`output_size=decoder.state_size`.
+        hparams (dict, optional): Hyperparameters. Missing
+            hyperparamerter will be set to default values. See
+            :meth:`default_hparams` for the hyperparameter sturcture and
+            default values.
+
+    The input to the connector can have arbitrary structure and size.
+    """
+
+    def __init__(self,
+                 output_size: OutputSize,
+                 hparams: HParamsType = None):
+        ConnectorBase.__init__(self, output_size, hparams)
+
+    @staticmethod
+    def default_hparams() -> dict:
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        .. code-block:: python
+
+            {
+                "activation_fn": "texar.core.layers.identity",
+                "name": "mlp_connector"
+            }
+
+        Here:
+
+        "activation_fn" : str or callable
+            The activation function applied to the outputs of the MLP
+            transformation layer. Can
+            be a function, or its name or module path.
+        "name" : str
+            Name of the connector.
+        """
+        return {
+            "activation_fn": "texar.core.layers.identity",
+            "name": "mlp_connector"
+        }
+
+    def forward(self, # type: ignore
+                inputs: Inputs
+                ) -> Any:
+        r"""Transforms inputs with an MLP layer and packs the results to have
+        the same structure as specified by :attr:`output_size`.
+
+        Args:
+            inputs: Input (structure of) tensors to be transformed. Must be a
+                ``Tensor`` of shape ``[batch_size, ...]`` or a (nested)
+                ``tuple`` of such Tensors. That is, the first dimension of
+                (each) ``Tensor`` must be the batch dimension.
+
+        Returns:
+            A ``Tensor`` or a (nested) ``tuple`` of Tensors of the
+            same structure of :attr:`output_size`.
+        """
+        activation_fn = layers.get_activation_fn(self.hparams.activation_fn)
+
+        output, linear_layer = _mlp_transform(
+            inputs, self._output_size, activation_fn)
+        self._linear_layers.append(linear_layer)
```

Comment:
This is incorrect in several ways:
1. Submodules and parameters must be added in `__init__` method, because users can call `.to` after constructing models to move module parameters to appropriate devices.
2. Current impl. will result in a new linear layer to be created every time `forward` is called. This is clearly incorrect. Constructor of `MLPTransformConnector` should take the final dimension size of input tensors as an argument.